### PR TITLE
Add rewriter to transform attributes into methods

### PR DESCRIPTION
**Pull request by amomchilov**
Part of https://github.com/Shopify/tapioca/issues/1918

This PR introduces a new rewriter which transforms `attr_reader`/`attr_writer`/`attr_accessor` RBIs into methods with the corresponding sigs. This has two benefits:

1. It naturally solves Shopify/tapioca#1709. If Tapioca only has to deal with methods (which can already be merged), then we don't have to write bespoke logic to merge attributes with methods, or methods with attributes.
2. It'll leave less work for Sorbet, which internally rewrites attributes into methods. We can do this on its behalf, once up-front, rather than on every parse of the RBIs.

### CODE REVIEW

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      )
+
+      [m]
+    end
+  end
+
+  class AttrWriter
+    sig { returns(T::Array[Method]) }
+    def convert_to_methods
+      raise "How can there be multiple?" unless sigs.one?
```
[the remark this answers is not in the document]
Comment (amomchilov) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb]: Not for attributes, though. [Sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20%7B%20returns%28Integer%29%20%7D%0Asig%20%7B%20returns%28Integer%29%20%7D%0Aattr_reader%20%3Afoo%0A)

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb]: In `.rbi` files only. With sorbet.run you're testing with a `.rb` file.

**Comment by amomchilov** [comment]
Spoke offline. Conclusions:

* Q1: Be flexible, and accept both `.void` and `.returns(Type)`. Reevaluate if Sorbet changes its behaviour to prefer one or the other
* Q2: Always emit `.void`.

File: lib/rbi/rewriters/replace_attributes_with_methods.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (amomchilov) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb]: Indeed, looks like it allows it, but the revealed type is whichever one you listed first. I think we should warn in that case, it likely wasn't intended.

[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb]: We currently do not do any linting / warning inside `rbi`, it's up to the client to make sure that what is produced is correct.

This could be a warning in Tapioca when pulling the gem's hand-written annotations. Or it could be done by Sorbet.

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        params(
+          name: String,
+          sig: T.nilable(Sig),
+          attribute_type: T.nilable(String),
+          visibility: Visibility,
+          loc: T.nilable(Loc),
+          comments: T::Array[Comment],
+        ).returns(Method)
+      end
+      def create_setter_method(name, sig, attribute_type, visibility, loc, comments) # rubocop:disable Metrics/ParameterLists
```
Comment (amomchilov) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:119]: 6 params, but the upper limit is 5. Worth overriding?

File: test/rbi/rewriters/replace_attributes_with_methods_test.rb [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        def a=(a); end
+
+        # Lorum ipsum...
+        sig { params(b: Integer).void }
+        def b=(b); end
+
+        # Lorum ipsum...
+        sig { params(c: Integer).void }
+        def c=(c); end
+      RBI
```
Comment (amomchilov) [inline test/rbi/rewriters/replace_attributes_with_methods_test.rb:162]: On second thought, I think it might be better to group by property name, rather than by reader vs writer, i.e.:

```ruby
        sig { returns(Integer) }
        def a; end

        sig { params(a: Integer).void }
        def a=(a); end

        sig { returns(Integer) }
        def b; end

        sig { params(b: Integer).void }
        def b=(b); end

        sig { returns(Integer) }
        def c; end

        sig { params(c: Integer).void }
        def c=(c); end
```

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RBI
+  module TempHelpers
+    extend T::Sig
+  end
+
+  module Rewriters
+    class ReplaceAttributesWithMethods < Visitor
```
Comment (amomchilov) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:10]: Name's a bit wordy, and "Replace" is kind of implied by "Rewriters::".

How about:

```suggestion
    class AttrToMethods < Visitor
```

File: lib/rbi/model.rb [hunk 5]
```diff
@@ -42,6 +42,16 @@ def replace(node)
       self.parent_tree = nil
     end
 
+    sig { params(nodes: T::Enumerable[Node]).void }
+    def replace_with_multiple(nodes)
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/model.rb:46]: It seems very niche. Let's keep it in the rewriter, we can promote it if we find another use case for it.

File: lib/rbi/model.rb [hunk 6]
```diff
@@ -42,6 +42,16 @@ def replace(node)
...
+      tree = parent_tree
+      raise unless tree
+
+      # Does this work?
```
Comment (Morriar) [inline lib/rbi/model.rb:50]: Did you find a way to answer this question?

File: lib/rbi/model.rb [hunk 7]
```diff
@@ -1153,6 +1163,43 @@ def initialize(
...
+        is_abstract: T::Boolean,
+        is_override: T::Boolean,
+        is_overridable: T::Boolean,
+        is_final: T::Boolean,
+        type_params: T::Array[String],
+        checked: T.nilable(Symbol),
+        loc: T.nilable(Loc),
+      ).returns(Sig)
+    end
+    def new_with(
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/model.rb:1179]: What's the benefits over calling `Sig.new` directly?

File: lib/rbi/model.rb [hunk 8]
```diff
@@ -1171,6 +1218,15 @@ def ==(other)
         is_override == other.is_override && is_overridable == other.is_overridable && is_final == other.is_final &&
         type_params == other.type_params && checked == other.checked
     end
+
+    sig { override.returns(String) }
+    def inspect
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/model.rb:1223]: Let's not override the default `inspect`, you can call `node.string` for the same result

File: lib/rbi/rewriters/replace_attributes_with_methods.rb (hunk 4 above)
Comment (Morriar) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb:10]: Yeah I like it more, other rewriters always use an action verb 👍

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      extend T::Sig
+      include TempHelpers
```
Comment (Morriar) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:12]:
```suggestion

      include TempHelpers
```

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+      sig { override.params(node: T.nilable(Node)).void }
+      def visit(node)
+        return unless node
```
Comment (Morriar) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:16]:
```suggestion
        return unless node.is_a?(Tree)
```

so you can remove the case below

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        writers = attr.names.map do |name|
+          create_setter_method(name.to_s, sig, attribute_type, attr.visibility, attr.loc, attr.comments)
+        end
+
+        readers + writers
+      end
+
+      sig { params(attr: Attr).returns([T.nilable(Sig), T.nilable(String)]) }
+      def parse_sig_of(attr)
+        raise "Attributes cannot have more than 1 sig" if 1 < attr.sigs.count
```
Comment (Morriar) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:76]: I would rather not have RBI do this check, as far as we know this is perfectly valid RBI syntax, the semantics of it should be left to Tapioca/Sorbet

File: lib/rbi/rewriters/replace_attributes_with_methods.rb [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+            child.replace_with_multiple(new_methods)
+          end
+        end
+      end
+
+      private
+
+      sig { params(attr: Attr).returns(T::Array[Method]) }
+      def convert_to_methods(attr)
+        sig, attribute_type = parse_sig_of(attr)
```
Comment (Morriar) [inline lib/rbi/rewriters/replace_attributes_with_methods.rb:35]: We could move this to where we actually create the resulting methods and let them `translate` the sig rather than passing it around.

* going from `attr_reader :a` to `def a; end` shouldn't require any sig translation
* going from `attr_writer :a` to `def a=(val); end` may require a sig translation
* going from `attr_accessor :a` requires a sig translation only for the writer

File: lib/rbi/model.rb (hunk 6 above)
Comment (amomchilov) [reply lib/rbi/model.rb:50]: Yup, works :)

File: lib/rbi/rewriters/replace_attributes_with_methods.rb (hunk 9 above)
Comment (amomchilov) [reply lib/rbi/rewriters/replace_attributes_with_methods.rb]: Ah, these were indeed meant to be temporary while I refactored stuff around. They'll be gone shortly.

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        case node
+        when Tree
+          node.nodes.dup.each do |child|
+            visit(child)
+            next unless (attr = child).is_a?(Attr)
+
+            new_methods = convert_to_methods(attr)
+
+            replace(child, with: new_methods)
+          end
```
Comment (Morriar) [inline lib/rbi/rewriters/attr_to_methods.rb:22]: Can't we simplify this actually?

```suggestion
        return unless node

        case node
        when Attr
          new_methods = convert_to_methods(attr)
          replace(child, with: new_methods)
        when Tree
          visit_all(node.nodes)
        end
```

Comment (Morriar) [reply lib/rbi/rewriters/attr_to_methods.rb:22]: You could even hoist up the logic from `convert_to_methods`?

File: lib/rbi/model.rb [hunk 14]
```diff
@@ -380,6 +380,9 @@ def initialize(name, names, visibility: Public.new, sigs: [], loc: nil, comments
 
     sig { abstract.returns(T::Array[String]) }
     def fully_qualified_names; end
+
+    sig { abstract.returns(T::Array[Method]) }
+    def convert_to_methods; end
```
Comment (Morriar) [inline lib/rbi/model.rb:385]: Let's move this in lib/rbi/rewriters/attr_to_methods.rb with the other methods

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+module RBI
+  module Rewriters
+    class AttrToMethods < Visitor
+      extend T::Sig
+
+      sig { override.params(node: T.nilable(Node)).void }
+      def visit(node)
+        case node
+        when Tree
+          visit_all(node.nodes)
```
[the remark this answers is not in the document]
Comment (Morriar) [reply lib/rbi/rewriters/attr_to_methods.rb:13]: I think you need a test iterating over multiple nodes:

```rb
class Foo
  def foo; end
  attr_accessor :bar
  def baz; end
end
```

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+        when Attr
+          new_methods = node.convert_to_methods
+          replace(node, with: new_methods)
```
Comment (Morriar) [inline lib/rbi/rewriters/attr_to_methods.rb:17]:
```suggestion
          replace(node, with: node.convert_to_methods)
```

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        end
+      end
+
+      private
+
+      sig { params(node: Node, with: T::Array[Node]).void }
+      def replace(node, with:)
+        tree = node.parent_tree
+        raise unless tree
```
Comment (Morriar) [inline lib/rbi/rewriters/attr_to_methods.rb:26]: Let's raise a `AttrToMethods::Error`

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+      visitor.visit(self)
+    end
+  end
+
+  class Attr
+    private
+
+    sig(:final) { returns([T.nilable(Sig), T.nilable(String)]) }
+    def parse_sig
+      raise "Attributes cannot have more than 1 sig" if 1 < sigs.count
```
Comment (Morriar) [inline lib/rbi/rewriters/attr_to_methods.rb:49]: Let's raise a `AttrToMethods::Error`

File: lib/rbi/rewriters/attr_to_methods.rb [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+  module Rewriters
+    class AttrToMethods < Visitor
+      extend T::Sig
+
+      sig { override.params(node: T.nilable(Node)).void }
+      def visit(node)
+        case node
+        when Tree
+          visit_all(node.nodes.dup)
+
```
Comment (Morriar) [inline lib/rbi/rewriters/attr_to_methods.rb:29]:
```suggestion
```

**Review by Morriar — APPROVED** [review]
